### PR TITLE
Add OAuth2 package

### PR DESCRIPTION
**Pull request by emcfarlane**
Create a new oauth2 package for functionality to work with OAuth2. This package will be used to authenticate CLI clients with the BSR. The current functionality is limited to declaring types for use with the device authorization grants flow.

### CODE REVIEW

**Review by doriable — APPROVED** [review]
Looks reasonable to me -- up to you on whether you want to merge #3143 into this and then merge them together, or merge this first and then the other PR will automatically rebase.

**Review by mfridman — APPROVED** [review]

File: private/pkg/oauth2/oauth2.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	// ErrorCode is the error code.
+	ErrorCode ErrorCode `json:"error"`
+	// ErrorDescription is a human-readable description of the error. May be empty.
+	ErrorDescription string `json:"error_description,omitempty"`
+	// ErrorURI is a URI for the error. May be empty.
+	ErrorURI string `json:"error_uri,omitempty"`
+}
+
+// Error implements error.
+func (e *Error) Error() string {
```
Comment (bufdev) [inline private/pkg/oauth2/oauth2.go:52]: Should handle nil case

File: private/pkg/oauth2/oauth2.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	// ErrorCodeUnsupportedGrantType is an unsupported grant type error.
+	ErrorCodeUnsupportedGrantType ErrorCode = "unsupported_grant_type"
+	// ErrorCodeInvalidScope is an invalid scope error.
+	ErrorCodeInvalidScope ErrorCode = "invalid_scope"
+)
+
+// Error is an OAuth2 error.
+type Error struct {
+	// ErrorCode is the error code.
+	ErrorCode ErrorCode `json:"error"`
```
Comment (bufdev) [inline private/pkg/oauth2/oauth2.go:44]: Any reason you didn't add omitempty tag everywhere?

Comment (emcfarlane) [reply private/pkg/oauth2/oauth2.go:44]: Tags without omitempty are required.

**Review by bufdev — APPROVED** [review]

**Pull request merged**